File: front/src/models/DetalleFactura.ts
import { Factura } from "./Factura";
import { Producto } from "./Producto";

export class DetalleFactura {
  id: number;
  facturaId: Factura;
  cantidad: number;
  precioUnitario: number;
  subtotal: number;
  productoNombre?: string; // <-- Agregado para mostrar el nombre del producto

  constructor(
    id: number,
    facturaId: Factura,
    cantidad: number,
    precioUnitario: number,
    subtotal: number,
    productoNombre?: string // <-- Agregado
  ) {
    this.id = id;
    this.facturaId = facturaId;
    this.cantidad = cantidad;
    this.precioUnitario = precioUnitario;
    this.subtotal = subtotal;
    this.productoNombre = productoNombre; // <-- Agregado
  }
}

File: Back/src/main/java/com/BackPM/BackPM/models/Detalle_Factura.java
package com.BackPM.BackPM.models;

import jakarta.persistence.*;
import java.math.BigDecimal;

@Entity
@Table(name = "detalle_factura")
public class Detalle_Factura extends ABaseEntity {

    @ManyToOne(fetch = FetchType.EAGER) // Optimización de rendimiento con carga perezosa
    @JoinColumn(name = "id_factura", nullable = false, referencedColumnName = "id")
    private Factura factura;

    @Column(name = "cantidad", nullable = false)
    private Integer cantidad;

    @Column(name = "precio_unitario", nullable = false, precision = 10, scale = 2)
    private BigDecimal precioUnitario;

    @Column(name = "subtotal", nullable = false, precision = 12, scale = 2)
    private BigDecimal subtotal;

    // Getters y Setters
    public Factura getFactura() {
        return factura;
    }

    public void setFactura(Factura factura) {
        this.factura = factura;
    }

    public Integer getCantidad() {
        return cantidad;
    }

    public void setCantidad(Integer cantidad) {
        this.cantidad = cantidad;
        calcularSubtotal();
    }

    public BigDecimal getPrecioUnitario() {
        return precioUnitario;
    }

    public void setPrecioUnitario(BigDecimal precioUnitario) {
        this.precioUnitario = precioUnitario;
        calcularSubtotal();
    }

    public BigDecimal getSubtotal() {
        return subtotal;
    }

    public void setSubtotal(BigDecimal subtotal) {
        this.subtotal = subtotal;
    }

    // Método para calcular el subtotal
    public void calcularSubtotal() {
        if (this.cantidad != null && this.precioUnitario != null) {
            this.subtotal = this.precioUnitario.multiply(BigDecimal.valueOf(this.cantidad));
        } else {
            this.subtotal = BigDecimal.ZERO;
        }
    }
}
